indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/TaskMonitor.java

```diff
@@ -265,6 +263,22 @@ private void retry(String subTaskSpecId, MonitorEntry monitorEntry, TaskStatusPl
     }
   }
 
+  private T submitTask(SubTaskSpec<T> spec, int numAttempts)
+  {
+    T task = spec.newSubTask(numAttempts);
+    try {
+      indexingServiceClient.runTask(task);
```

Comment:
Will this approach be able to retry immediately, or will it have to exhaust retries in the `indexingServiceClient` first? (The loop in DruidLeaderClient)

Ideally, this detects a problem on the first submission, and doesn't need to exhaust retries before it moves on to trying the backwards-compatible type.

Could you check it, if you haven't already?

Reply:
This won't retry immediately but will exhaust retries in DruidLeaderClient. Yes, ideally it should be able to detect the problem before it retries, but I'm not sure whether that refactoring is worth to do because 1) it's not easy to teach the logic of the caller to DruidLeaderClient, 2) it will happen only during a particular type of rolling update, 3) and retries won't take much time compared to the total indexing time.